README.md
# Express Form2

Express Form provides data filtering and validation as route middleware to your Express applications.



## Requires

Checkout `package.json` for dependencies.



## Installation

Install express-form2 through npm

    npm install express-form2



## API

- field( name ).isArray( msg );
- field( name ).isAlpha( msg );
- field( name ).isAlphanumeric( msg );
- field( name ).isDate( msg );
- field( name ).isDecimal( msg );
- field( name ).isFloat( msg );
- field( name ).isNumeric( msg );
- field( name ).isInt( msg );
- field( name ).isNull( msg );
- field( name ).notNull( msg );
- field( name ).isCreditCard( msg );
- field( name ).isEmail( msg );
- field( name ).isIP( msg );
- field( name ).isLowercase( msg );
- field( name ).isUppercase( msg );
- field( name ).isUrl( msg );
- field( name ).isUUID( version );
- field( name ).required( msg );
- field( name ).notEmpty( msg );
- field( name ).contains( val, msg );
- field( name ).notContains( val, msg );
- field( name ).is( pattern, modifiers, msg );
- field( name ).regex( pattern, modifiers, msg );
- field( name ).not( pattern, modifiers, msg );
- field( name ).notRegex( pattern, modifiers, msg );
- field( name ).isIn( opt, msg );
- field( name ).notIn( opt, msg );
- field( name ).equals( eql, msg );
- field( name ).isAfter( date, msg );
- field( name ).isBefore( date, msg );
- field( name ).len( min, max, msg );
- field( name ).max( val, msg );
- field( name ).min( val, msg );
- field( name ).maxLength( val, msg );
- field( name ).minLength( val, msg );
- field( name ).custom( func, msg );
- field( name ).ifNull( replace );
- field( name ).entityDecode();
- field( name ).entityEncode();
- field( name ).ltrim( chars );
- field( name ).rtrim( chars );
- field( name ).toFloat();
- field( name ).toInt();
- field( name ).toBoolean();
- field( name ).toBooleanStrict();
- field( name ).trim( chars );
- field( name ).xss( opt );
- field( name ).toArray();
- field( name ).toUpper();
- field( name ).toLower();
- field( name ).truncate( len );



## Usage

### Example

    var form    = require( 'express-form2' );
    var field   = form.field;
    var express = require( 'express' );
    var app     = express.createServer();

    app.configure( function(){
      app.use( express.bodyDecoder());
      app.use( app.router );
    });

    app.post(

      // Route
      '/user',

      // Form filter and validation middleware
      form(
        field( 'username' ).trim().required().is( /^[a-z]+$/ ),
        field( 'password' ).trim().required().is( /^[0-9]+$/ ),
        field( 'email' ).trim().isEmail()
      ),

      // Express request-handler now receives filtered and validated data
      function( req, res ){
        if( !req.form.isValid ){
          // Handle errors
          console.log( req.form.errors );
        }else{
          // Or, use filtered form data from the form object:
          console.log( 'Username:', req.form.username );
          console.log( 'Password:', req.form.password );
          console.log( 'Email:', req.form.email );
        }
      }
    );



### Module

The Express Form **module** returns an Express [Route Middleware](http://expressjs.com/guide.html#Route-Middleware) function. You specify filtering and validation by passing filters and validators as arguments to the main module function. For example:

    var form = require( 'express-form2' );

    app.post( '/user',

      // Express Form Route Middleware: trims whitespace off of
      // the `username` field.
      form( form.field( 'username' ).trim()),

      // standard Express handler
      function ( req, res ){
        // ...
      }
    );



#### http.ServerRequest.prototype.form

Express Form adds a `form` object with various properties to the request.

    isValid -> Boolean

    errors  -> Array

    flashErrors(name) -> undefined

        Flashes all errors. Configurable, enabled by default.

    getErrors(name) -> Array or Object if no name given
    - fieldname (String): The name of the field

        Gets all errors for the field with the given name.

        You can also call this method with no parameters to get a map of errors for all of the fields.

    Example request handler:

    function ( req, res ){
      if( !req.form.isValid ){
        console.log( req.form.errors );
        console.log( req.form.getErrors( 'username' ));
        console.log( req.form.getErrors());
      }
    }



#### Options / Configuration
Express Form has various configuration options, but aims for sensible defaults for a typical Express application.

    form.configure(options) -> self
    - options (Object): An object with configuration options.

    flashErrors (Boolean): If validation errors should be automatically passed to Express’ flash() method. Default: true.

    autoLocals (Boolean): If field values from Express’ request.body should be passed into Express’ response.locals object. This is helpful when a form is invalid an you want to repopulate the form elements with their submitted values. Default: true.

    Note: if a field name dash-separated, the name used for the locals object will be in camelCase.

    dataSources (Array): An array of Express request properties to use as data sources when filtering and validating data. Default: ["body", "query", "params"].

    autoTrim (Boolean): If true, all fields will be automatically trimmed. Default: false.

    passThrough (Boolean): If true, all data sources will be merged with `req.form`. Default: false.

    errorCodeOnly (Boolean): If true the error msg will only return error code.
    Default: false.



### Fields

The `field` property of the module creates a filter/validator object tied to a specific field.

    var form  = require( 'express-form2' );
    var field = form.field;

    field( fieldname[, label ]);

You can access nested properties with either dot or square-bracket notation.

    field( 'post.content' ).minLength( 50 );
    field( 'post[ user ][ id ]' ).isInt();
    field( 'post.super.nested.property' ).required();

Simply specifying a property like this, makes sure it exists. So, even if `req.body.post` was undefined, `req.form.post.content` would be defined. This helps avoid any unwanted errors in your code.

The API is chainable, so you can keep calling filter/validator methods one after the other:

    field( 'username' ).trim().toLower().truncate( 5 ).required().isAlphanumeric();

## Sanitization / Filters

### field( name ).ifNull( replace );
    ', undefined and null get replaced by `replacement`

### field( name ).entityDecode();
    Decode HTML entities

### field( name ).entityEncode();
    Encodes HTML entities

### field( name ).ltrim( chars );
### field( name ).rtrim( chars );
### field( name ).toFloat();
    Number

### field( name ).toInt();
    Number, rounded down

### field( name ).toBoolean();
    Boolean from truthy and falsy values. True unless str = '0', 'false', or str.length == 0

### field( name ).toBooleanStrict();
    Only true, 'true', 1 and '1' are `true`. False unless str = '1' or 'true'

### field( name ).trim( chars );
    Trim optional `chars`, default is to trim whitespace ( \r\n\t\s )

### field( name ).xss( opt );
    Remove common XSS attack vectors from text ( true from images )

### field( name ).toArray();
    Using the toArray() flag means that field always gives an array. If the field value is an array, but there is no flag, then the first value in that array is used instead.

        This means that you don't have to worry about unexpected post data that might break your code. Eg/ when you call an array method on what is actually a string.

        field( 'project.users' ).toArray(),
        // undefined => [], '' => [], 'q' => [ 'q' ], [ 'a', 'b' ] => [ 'a', 'b' ]

        field( 'project.block' ),
        // project.block : [ 'a', 'b' ] => 'a'. No 'toArray()', so only first value used.

        In addition, any other methods called with the array method, are applied to every value within the array.

        field( 'post.users' ).toArray().toUpper()
        // post.users : [ 'one', 'two', 'three' ] => [ 'ONE', 'TWO', 'THREE' ]

### field( name ).toUpper();
### field( name ).toLower();
### field( name ).truncate( len );
    Chops value at (length - 3), appends `...`

## Validations

**Validation messages**: each validator has its own default validation message. These can easily be overridden at runtime by passing a custom validation message to the validator. The custom message is always the **last** argument passed to the validator.

Use '%s' in the message to have the field name or label printed in the message:

    field( 'username' ).required()
    // -> 'username is required'

    field( 'username' ).required( 'What is your %s?' )
    // -> 'What is your username?'

    field( 'username', 'Username' ).required( 'What is your %s?' )
    // -> 'What is your Username?'

> By type

### field( name ).isArray( msg );
### field( name ).isAlpha( msg );
### field( name ).isAlphanumeric( msg );
### field( name ).isDate( msg );
    Uses Date.parse() - regex is probably a better choice

### field( name ).isDecimal( msg );
### field( name ).isFloat( msg );
    Alias for isDecimal

### field( name ).isNumeric( msg );
### field( name ).isInt( msg );
    isNumeric accepts zero padded numbers, e.g. '001', isInt doesn't

### field( name ).isNull( msg );
### field( name ).notNull( msg );

> By format

### field( name ).isCreditCard( msg );
    Will work against Visa, MasterCard, American Express, Discover, Diners Club, and JCB card numbering formats

### field( name ).isEmail( msg );
### field( name ).isIP( msg );
### field( name ).isLowercase( msg );
### field( name ).isUppercase( msg );
### field( name ).isUrl( msg );
    Accepts http, https, ftp

### field( name ).isUUID( version );
    Version can be 3 or 4 or empty, see http://en.wikipedia.field( name ).org/wiki/Universally_unique_identifier

> By content

### field( name ).required( msg );
    Checks that the field is present in form data, and has a value.

### field( name ).notEmpty( msg );
    Checks if the value is not just whitespace.

### field( name ).contains( val, msg );
    - val (String): The value to test for.

      Checks if the field contains `value`.

### field( name ).notContains( val, msg );
    - value (String): A value that should not exist in the field.

        Checks if the field does NOT contain `value`.

### field( name ).is( pattern, modifiers, msg );
    Alias for regex()

### field( name ).regex( pattern, modifiers, msg );
    regex( pattern[, modifiers[, message ]])
    - pattern (RegExp|String): RegExp (with flags) or String pattern.
    - modifiers (String): Optional, and only if `pattern` is a String.
    - message (String): Optional validation message.

        alias: is

        Checks that the value matches the given regular expression.

        Example:

        field( 'username' ).is( '[a-z]', 'i', 'Only letters are valid in %s' );
        field( 'username' ).is( /[a-z]/i, 'Only letters are valid in %s' );

### field( name ).not( pattern, modifiers, msg );
    Alias for notRegex()

### field( name ).notRegex( pattern, modifiers, msg );
    notRegex( pattern[, modifiers[, message ]])
    - pattern (RegExp|String): RegExp (with flags) or String pattern.
    - modifiers (String): Optional, and only if `pattern` is a String.
    - message (String): Optional validation message.

        alias: not

        Checks that the value does NOT match the given regular expression.

        Example:

        field( 'username' ).not( '[a-z]', 'i', 'Letters are not valid in %s' );
        field( 'username' ).not( /[a-z]/i, 'Letters are not valid in %s' );

### field( name ).isIn( opt, msg );
    Accepts an array or a string

### field( name ).notIn( opt, msg );
    Accepts an array or a string

### field( name ).equals( eql, msg );
    equals( value [, message])
    - value (String): A value that should match the field value OR a fieldname
                      token to match another field, ie, `field::password`.

        Compares the field to `value`.

        Example:
        field( 'username' ).equals( 'admin' );

        field( 'password' ).is( /^\w{6,20}$/ );
        field( 'password_confirmation' ).equals( 'field::password' );

### field( name ).isAfter( date, msg );
    Argument is optional and defaults to today

### field( name ).isBefore( date, msg );
    Argument is optional and defaults to today

### field( name ).len( min, max, msg );
    For both str & arr, max is optional

### field( name ).max( val, msg );
### field( name ).min( val, msg );
### field( name ).maxLength( val, msg );
### field( name ).minLength( val, msg );
### field( name ).custom( func, msg );
    custom( function[, message ])
    - function (Function): A custom filter or validation function.

        This method can be utilised as either a filter or validator method.

        If the function throws an error, then an error is added to the form. (If `message` is not provided, the thrown error message is used.)

        If the function returns a value, then it is considered a filter method, with the field then becoming the returned value.

        If the function returns undefined, then the method has no effect on the field.

        Examples:

        If the `name` field has a value of 'hello there', this would
        transform it to 'hello-there'.

        field( 'name' ).custom( function ( value ){
          return value.replace( /\s+/g, '-' );
        });

        Throws an error if `username` field does not have value 'admin'.

        field( 'username' ).custom( function ( value ){
          if( value !== 'admin' ){
            throw new Error( '%s must be `admin`.' );
          }
        });

## Error Code

- By type
    - 10 : isArray
    - 11 : isAlpha
    - 12 : isAlphanumeric
    - 13 : isDate
    - 14 : isDecimal
    - 14 : isFloat
    - 15 : isNumeric
    - 16 : isInt
    - 17 : isNull
    - 18 : notNull

- By format
    - 20 : isCreditCard
    - 21 : isEmail
    - 22 : isIP
    - 23 : isLowercase
    - 24 : isUppercase
    - 25 : isUrl
    - 26 : isUUID

- By content
    - 30  : required
    - 31  : notEmpty
    - 32  : contains
    - 33  : notContains
    - 34  : is
    - 34  : regex
    - 35  : not
    - 35  : notRegex
    - 36  : isIn
    - 37  : notIn
    - 38  : equals
    - 39  : isAfter
    - 310 : isBefore
    - 311 : len
    - 312 : max
    - 312 : maxLength
    - 313 : min
    - 313 : minLength
    - 0   : custom



## Runing Tests

    npm install expresso -g
    cd /path/to/the/project/root
    expresso



## Credits

Currently, Express Form uses many of the validation and filtering functions provided by Chris O'Hara's [node-validator]( https://github.com/chriso/node-validator ).



## License

(The MIT License)

Copyright (c) 2011 dreamerslab &lt;ben@dreamerslab.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software' ), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
# History

## 1.5.0 / 2015-09-29

- [update packages] inflection->1.7.1, node.extend->1.1.5
- [bug fix] Validator `isFloat` now works properly



## 1.4.4 / 2014-05-12

- [update packages] inflection->1.3.5, node.extend->1.0.10, validator->2.1.0



## 1.4.3 / 2013-09-11

- [update packages] node.extend->1.0.8, validator->1.5.1,



## 1.4.2 / 2013-06-29

- [bug fix] Return when the param is null



## 1.4.1 / 2013-04-20

- [bug fix] Typo `ture` -> `true`



## 1.4.0 / 2013-04-20

- [bug fix] Validator `is` only take string now
- [new feature] Add `isBool` & `isBoolean`
- [refactoring] Better `utils.typeof`



## 1.3.2 / 2013-04-14

- [bug fix] Fix the same fucking bug



## 1.3.1 / 2013-04-14

- [update packages] node.extend->1.0.3
- [bug fix] Make sure `isArray` & `toArray` work properly



## 1.3.0 / 2013-04-12

- [update packages] inflection->1.2.5, express->2.5.11, node.extend->1.0.2, validator->0.5.0
- [refactoring] Array value has to call `isArray` first



## 1.2.0 / 2013-01-02

- [bug fix] For array elements stay as array elements even without calling `toArray`



## 1.1.5 / 2012-07-17

- [bug fix] Undefined value for mix & max length



## 1.1.4 / 2012-06-26

- [update packages] inflection->1.2.1, express->2.5.10, validator->0.4.9



## 1.1.3 / 2012-05-30

- [bug fix] Skip `undefined` value no matter what



## 1.1.2 / 2012-05-01

- [bug fix] Skip `undefined` value if its not required



## 1.1.1 / 2012-05-01

- [refactoring] Returns error only if the field is required for isArray
- [refactoring] Use default structure to show error msgs



## 1.1.0 / 2012-05-01

- [refactoring] Remove default value for none array prop



## 1.0.0 / 2012-05-01

- Initial release


test/form.test.js
var assert = require("assert"),
    form = require("../index"),
    validate = form.validate;

module.exports = {
  'form : isValid': function() {
    // Failure.
    var request = { body: { field: "fail" }};
    form(validate("field").isEmail())(request, {});
    assert.strictEqual(request.form.isValid, false);

    // Success
    var request = { body: { field: "me@dandean.com" }};
    form(validate("field").isEmail())(request, {});
    assert.strictEqual(request.form.isValid, true);

    // assert["throws"](function() {
    //   request.form.isValid = false;
    // });

    assert.strictEqual(request.form.isValid, true);
  },

  'form : getErrors': function() {
    var request = {
      body: {
        field0: "win",
        field1: "fail",
        field2: "fail",
        field3: "fail"
      }
    };

    form(
      validate("field0").equals("win"),
      validate("field1").isEmail(),
      validate("field2").isEmail().isUrl(),
      validate("field3").isEmail().isUrl().isIP()
    )(request, {});

    assert.equal(request.form.isValid, false);
    assert.equal(request.form.errors.length, 6);

    assert.equal(request.form.getErrors("field0").length, 0);
    assert.equal(request.form.getErrors("field1").length, 1);
    assert.equal(request.form.getErrors("field2").length, 2);
    assert.equal(request.form.getErrors("field3").length, 3);
  },

  'form : configure : dataSources': function() {
    form.configure({ dataSources: 'other' });

    var request = { other: { field: "me@dandean.com" }};
    form(validate("field").isEmail())(request, {});
    assert.strictEqual(request.form.isValid, true);
    assert.equal(request.form.field, "me@dandean.com");

    form.configure({ dataSources: ['body', "query", "params"] });
  }
};


test/more.test.js
var assert = require( 'assert' );
var form   = require( '../index' );
var field  = form.field;

module.exports = {

  'field : arrays' : function (){
    // Array transformations.
    var request = {
      body     : {
        field1 : '',
        field2 : 'Hello!',
        field3 : [ 'Alpacas?', 'Llamas!!?', 'Vicunas!', 'Guanacos!!!' ]
      }
    };

    form(
      field( 'fieldx' ).toArray(),
      field( 'field1' ).toArray(),
      field( 'field2' ).toArray(),
      field( 'field3' ).toArray()
    )( request, {});

    assert.strictEqual( Array.isArray( request.form.fieldx ), true );
    assert.strictEqual( request.form.fieldx.length, 0 );
    assert.strictEqual( Array.isArray( request.form.field1 ), true );
    assert.strictEqual( request.form.field1.length, 0 );
    assert.strictEqual( request.form.field2[ 0 ], 'Hello!' );
    assert.strictEqual( request.form.field2.length, 1 );
    assert.strictEqual( request.form.field3[ 0 ], 'Alpacas?' );
    assert.strictEqual( request.form.field3[ 1 ], 'Llamas!!?' );
    assert.strictEqual( request.form.field3[ 2 ], 'Vicunas!' );
    assert.strictEqual( request.form.field3[ 3 ], 'Guanacos!!!' );
    assert.strictEqual( request.form.field3.length, 4 );

    // No array flag!
    var request = { body : { field : [ 'red', 'blue' ]}};

    form( field( 'field' ))( request, {});
    assert.strictEqual( request.form.field[ 0 ], 'red' );
    assert.strictEqual( request.form.field[ 1 ], 'blue' );

    // Iterate and filter array.
    var request = { body : { field : [ 'david', 'stephen', 'greg' ]}};

    form( field( 'field' ).toArray().toUpper())( request, {});
    assert.strictEqual( request.form.field[ 0 ], 'DAVID' );
    assert.strictEqual( request.form.field[ 1 ], 'STEPHEN' );
    assert.strictEqual( request.form.field[ 2 ], 'GREG' );
    assert.strictEqual( request.form.field.length, 3 );

    // Iterate and validate array
    var request = { body : { field : [ 1, 2, 'f' ]}};

    form( field( 'field' ).toArray().isInt())( request, {});
    assert.equal( request.form.errors.length, 1 );
    assert.equal( request.form.errors[ 0 ], 'field is not an integer' );
  },
  'field : nesting' : function (){
    // Nesting with dot notation
    var request = {
      body          : {
        field       : {
          nest      : 'wow',
          child     : '4',
          gb        : {
            a       : 'a',
            b       : 'aaaa',
            c       : {
              fruit : 'deeper',
              must  : {
                go  : 'deeperrrr'
              }
            }
          }
        }
      }
    };

    form(
      field( 'field.nest' ).toUpper(),
      field( 'field.child' ).toUpper(),
      field( 'field.gb.a' ).toUpper(),
      field( 'field.gb.b' ).toUpper(),
      field( 'field.gb.c.fruit' ).toUpper(),
      field( 'field.gb.c.must.go' ).toUpper()
    )( request, {});

    assert.strictEqual( request.form.field.nest, 'WOW' );
    assert.strictEqual( request.form.field.child, '4' );
    assert.strictEqual( request.form.field.gb.a, 'A' );
    assert.strictEqual( request.form.field.gb.b, 'AAAA' );
    assert.strictEqual( request.form.field.gb.c.fruit, 'DEEPER' );
    assert.strictEqual( request.form.field.gb.c.must.go, 'DEEPERRRR' );

    // Nesting with square-bracket notation
    var request = {
      body          : {
        field       : {
          nest      : 'wow',
          child     : '4',
          gb        : {
            a       : 'a',
            b       : 'aaaa',
            c       : {
              fruit : 'deeper',
              must  : {
                go  : 'deeperrrr'
              }
            }
          }
        }
      }
    };

    form(
      field( 'field[nest]' ).toUpper(),
      field( 'field[child]' ).toUpper(),
      field( 'field[gb][a]' ).toUpper(),
      field( 'field[gb][b]' ).toUpper(),
      field( 'field[gb][c][fruit]' ).toUpper(),
      field( 'field[gb][c][must][go]' ).toUpper()
    )( request, {});

    assert.strictEqual( request.form.field.nest, 'WOW' );
    assert.strictEqual( request.form.field.child, '4' );
    assert.strictEqual( request.form.field.gb.a, 'A' );
    assert.strictEqual( request.form.field.gb.b, 'AAAA' );
    assert.strictEqual( request.form.field.gb.c.fruit, 'DEEPER' );
    assert.strictEqual( request.form.field.gb.c.must.go, 'DEEPERRRR' );
  },

  'field : filter/validate combo ordering' : function (){
    // Can arrange filter and validate procs in any order.
    var request = {
      body     : {
        field1 : '    whatever    ',
        field2 : '    some thing     '
      }
    };

    form(
      field( 'field1' ).trim().toUpper().maxLength( 5 ),
      field( 'field2' ).minLength( 12 ).trim()
    )( request, {});

    assert.strictEqual( request.form.field1, 'WHATEVER' );
    assert.strictEqual( request.form.field2, 'some thing' );
    assert.equal( request.form.errors.length, 1 );
    assert.equal( request.form.errors[ 0 ], 'field1 is too long' );
  },

  'field : autoTrim' : function (){
    // Auto-trim declared fields.
    var request = { body : {  field : '    whatever    ' }};

    form.configure({ autoTrim : true });
    form( field( 'field' ))( request, {});
    assert.strictEqual( request.form.field, 'whatever' );
    form.configure({ autoTrim : false });
  },

  'field : passThrough' : function (){
    // request.form gets all values from sources.
    var request = {
      body     : {
        field1 : 'fdsa',
        field2 : 'asdf'
      }
    };

    form.configure({ passThrough : true });
    form( field( 'field1' ))( request, {});
    assert.strictEqual( request.form.field1, 'fdsa' );
    assert.strictEqual( request.form.field2, 'asdf' );

    // request.form only gets declared fields.
    var request = {
      body     : {
        field1 : 'fdsa',
        field2 : 'asdf'
      }
    };

    form.configure({ passThrough : false });
    form( field( 'field1' ))( request, {});
    assert.strictEqual( request.form.field1, 'fdsa' );
    assert.strictEqual( typeof request.form.field2, 'undefined' );
  },

  'form : getErrors() gives full map' : function(){
    var request = {
      body     : {
        field0 : 'win',
        field1 : 'fail',
        field2 : 'fail',
        field3 : 'fail'
      }
    };

    form(
      field( 'field0' ).equals( 'win' ),
      field( 'field1' ).isEmail(),
      field( 'field2' ).isEmail().isUrl(),
      field( 'field3' ).isEmail().isUrl().isIP()
    )( request, {});

    assert.equal( request.form.isValid, false );
    assert.equal( request.form.errors.length, 6 );
    assert.equal( typeof request.form.getErrors().field0, 'undefined' );
    assert.equal( request.form.getErrors().field1.length, 1 );
    assert.equal( request.form.getErrors().field2.length, 2 );
    assert.equal( request.form.getErrors().field3.length, 3 );
  }
};


lib/form.js
/*!
 * express-form2
 * Copyright(c) 2011 Ben Lin <ben@dreamerslab.com>
 * MIT Licensed
 *
 * @fileoverview
 *
 */

var extend     = require( 'node.extend' );
var inflection = require( 'inflection' );
var utils      = require( './utils' );
var Field      = require( './field' );
var msg        = require( './msg' );

function form (){
  var self  = this;
  var opt   = form._opt;
  var tasks = [].slice.call( arguments );

  return function ( req, res, next ){
    var map          = {};
    var flashed      = {};
    var mergedSource = {};

    opt.dataSources.forEach( function ( src ){
      extend( mergedSource, req[ src ]);
    });

    if( !req.form ) req.form = {};

    if( opt.passThrough ) req.form = extend( mergedSource );

    if( opt.autoLocals ){
      for( var prop in req.body ){
        if( !req.body.hasOwnProperty( prop )) continue;

        if( typeof res.locals === 'function' ){ // Express 2.0 Support
          res.local( inflection.camelize( prop ), req.body[ prop ]);
        }

        if( !res.locals ) res.locals = {};
        res.locals[ inflection.camelize( prop )] = req.body[ prop ];
      }
    }

    Object.defineProperties( req.form, {
      'errors' : {
        value      : [],
        enumerable : false
      },
      'getErrors' : {
        value : function ( name ){
          if( !name ) return map;

          return map[ name ] || [];
        },
        enumerable : false
      },
      'isValid' : {
        get : function (){
          return this.errors.length === 0;
        },
        enumerable : false
      },
      'flashErrors' : {
        value : function (){
          if( typeof req.flash !== 'function' ) return;

          this.errors.forEach( function ( error ){
            if( flashed[ error ]) return;

            flashed[ error ] = true;
            req.flash( 'error', error );
          });
        },
        enumerable : false
      }
    });

    tasks.forEach( function ( task ){
      var result = task.run( mergedSource, req.form, opt );

      if( !Array.isArray( result ) || !result.length ) return;

      var name = task.name;

      req.form.errors = req.form.errors || [];
      map[ name ]     = map[ name ] || [];

      result.forEach( function ( error ){
        req.form.errors.push( error );
        map[ name ].push( error );
      });
    });

    if( opt.flashErrors ) req.form.flashErrors();
    if( next ) next();
  };
};

form._opt = {
  dataSources   : [ 'body', 'query', 'params' ],
  autoTrim      : false,
  autoLocals    : true,
  passThrough   : false,
  flashErrors   : true,
  errorCodeOnly : false
};

form.field = function ( prop, label ){
  return new Field( prop, label, msg( form._opt.errorCodeOnly ));
};

form.configure = function ( opt ){
  var self = this;

  Object.keys( opt ).forEach( function ( key ){
    if( !Array.isArray( opt[ key ]) && key === 'dataSources' ){
      opt[ key ] = [ opt[ key ]];
    }

    self._opt[ key ] = opt[ key ];
  });

  return this;
};

form.filter = form.validate = form.field;

module.exports = form;


lib/field.js
var extend          = require( 'node.extend' );
var utils           = require( './utils' );
var filters         = require( './filters' );
var validators      = require( './validators' );
var validator       = require( 'validator' );
var FilterPrototype = validator.Filter.prototype;
var externalFilter  = new validator.Filter();

var Field = function ( prop, label, msg ){
  var self  = this;
  var tasks = [];

  label = label || prop;

  this.name       = prop;
  this.__required = false;
  this.__isArray  = false;
  this.msg        = msg;

  this.add = function ( func ){
    tasks.push( func );
    return this;
  };

  this.run_tasks = function ( val ){
    tasks.forEach( function ( rule ){
      var result = rule( val, self.src ); // Pass src for 'equals' rule.

      if( result === undefined || result === null ) return;
      if( result.valid ) return;
      if( result.error ){
        // If this field is not required and it doesn't have a val, ignore error.
        if( !utils.hasValue( val ) && !self.__required ) return;

        return self.errors.push( result.error.replace( '%s', label ));
      }

      val = result;
    });

    return val;
  };

  this.run = function ( src, form, opt ){
    var val = utils.get( prop, form ) || utils.get( prop, src );

    this.src            = src;
    this.errors         = [];
    this.__checkIsArray = this.__isArray;

    if( opt.autoTrim ){
      tasks.unshift( function ( val ){
        return utils.typeof( val ) === 'string' ?
          FilterPrototype.trim.apply( externalFilter.sanitize( val )) :
          val;
      });
    }

    if( this.__isArray ){
      if( this.__forceArray ){
        if( !utils.hasValue( val )) val = [];
        if( !Array.isArray( val ))  val = [ val ];

        val = val.map( this.run_tasks );
      }else{
        if( Array.isArray( val )){
          val = val.map( this.run_tasks );
        }else{
          this.__checkIsArray = false;
          val = this.run_tasks( val );
        }
      }
    }else{
      val = this.run_tasks( val );
    }

    utils.set( prop, form, val );

    if( this.errors.length ) return this.errors;
  };
};

extend( Field.prototype, filters );
extend( Field.prototype, validators );

module.exports = Field;


lib/msg.js
module.exports = function ( errorCodeOnly ){
  return errorCodeOnly ? {
    // type
    isArray        : '10',
    isAlpha        : '11',
    isAlphanumeric : '12',
    isDate         : '13',
    isDecimal      : '14',
    isFloat        : '14', // alias for isDecimal
    isNumeric      : '15',
    isInt          : '16',
    isNull         : '17',
    notNull        : '18',
    isBool         : '19',
    isBoolean      : '19',
    // format
    isCreditCard   : '20',
    isEmail        : '21',
    isIP           : '22',
    isLowercase    : '23',
    isUppercase    : '24',
    isUrl          : '25', // http, https, ftp
    isUUID         : '26',
    // content
    required       : '30',
    notEmpty       : '31',
    contains       : '32',
    notContains    : '33',
    is             : '34', // alias for regex
    regex          : '34',
    not            : '35', // alias for noregex
    notRegex       : '35',
    isIn           : '36', // array or string
    notIn          : '37', // array or string
    equals         : '38',
    isAfter        : '39', // date
    isBefore       : '310', // date
    len            : '311', // array or string, max is optional
    max            : '312',
    min            : '313',
    maxLength      : '312',
    minLength      : '313',
    custom         : '0'
  } : {
    // type
    isArray        : '%s is not an array',
    isAlpha        : '%s contains non-letter characters',
    isAlphanumeric : '%s contains non alpha-numeric characters',
    isDate         : '%s is not a valid date',
    isDecimal      : '%s is not a decimal',
    isFloat        : '%s is not a decimal', // alias for isDecimal
    isNumeric      : '%s is not a number',
    isInt          : '%s is not an integer',
    isNull         : '%s is not null',
    notNull        : '%s is null',
    isBool         : '%s is not boolean',
    isBoolean      : '%s is not boolean',
    // format
    isCreditCard   : '%s is not a credit card number',
    isEmail        : '%s is not an email address',
    isIP           : '%s is not an IP address',
    isLowercase    : '%s contains uppercase letters',
    isUppercase    : '%s contains lowercase letters',
    isUrl          : '%s is not an URL', // http, https, ftp
    isUUID         : '%s is not an UUID',
    // content
    required       : '%s is required',
    notEmpty       : '%s has no value or is only whitespace',
    contains       : '%s does not contain required characters',
    notContains    : '%s contains invalid characters',
    is             : '%s has invalid characters', // alias for regex
    regex          : '%s has invalid characters',
    not            : '%s has invalid characters', // alias for noregex
    notRegex       : '%s has invalid characters',
    isIn           : '%s has invalid characters', // array or string
    notIn          : '%s has invalid characters', // array or string
    equals         : '%s does not equal to %t',
    isAfter        : '%s is not a valid date', // date
    isBefore       : '%s is not a valid date', // date
    len            : '%s is not in range', // array or string, max is optional
    max            : '%s is invalid',
    min            : '%s is invalid',
    maxLength      : '%s is too long',
    minLength      : '%s is too short',
    custom         : '%s is invalid'
  }
};
